pkg/auth/providers/azure/azure_provider.go

```diff
@@ -400,6 +415,27 @@ func formAzureRedirectURL(config map[string]interface{}) string {
 	)
 }
 
+func extractAnnotations(config map[string]interface{}) map[string]string {
+	annotations := make(map[string]string)
+	metadata, ok := config["metadata"].(map[string]interface{})
+	if !ok {
+		return annotations
+	}
+	rawAnnotations, ok := metadata["annotations"].(map[string]interface{})
+	if !ok {
+		return annotations
+	}
+	for k, v := range rawAnnotations {
+		if stringValue, ok := v.(string); ok {
+			annotations[k] = stringValue
+		}
+	}
+	if _, ok := annotations[GraphEndpointMigratedAnnotation]; !ok {
```

Comment:
I don't think this log condition meets what I was thinking about in https://github.com/rancher/rancher/pull/38125/files#r913802917 and might actually do more harm than good. My main concern here was that you have several places in this function where you return early because you couldn't parse the object into an expected form (like when you look into `config["mateadata"]` or `metadata["annotations"]`). In these cases, you do not return an error but instead return an empty map. It's impossible to distinguish these cases from the cases where there were no annotations, and since there is no log statement, there's no way of telling what happened or why we didn't parse the value.

In addition, this new log statement seems to have awareness of why we are parsing the annotation (to look for the migratedAnnotation). That seems to violate separation of responsibilities - this function doesn't care why we want the annotations (and doesn't use that value at all outside of a log warning), so it shouldn't check for a specific value in the annotations.

In short:
- This still leaves room for "ghost errors" which are not logged.
- Checking for this specific value is out of scope for this function
- Since we expect this value to be missing on legacy configurations, you are going to have a lot of these log statements for a missing annotation

I would recommend one of the following:
- Log the failure at each early return point in the function. If you also want to log in another function (such as isConfigDeprecated - not recommended) that we couldn't find the annotation that's ok, but that should be `trace`/`info` at most, not `warn`. 
- Return an error from this function at the early return points, and handle in the calling function (i.e. formAzureRedirectURL - which can choose to ignore and log `couldn't get annotations, error %s, will use old code`)

Reply:
Very good point, will make sure the logging happens before all returns from the function.